_posts/2020-08-01-data-science-meets-devops.md

```diff
@@ -0,0 +1,262 @@
+platform-aws | 0.8 | 0.5
+platform-gcp | 0.8 | 0.6
+
+<figcaption><strong>Table 1:</strong> Evaluation metrics for various Kubeflow labels.</figcaption>
+
+Given the rate at which new issues are arriving, retraining our model periodically became a priority. We believe continuously retraining and deploying our model to leverage this new data is critical to maintaining the efficacy of our models.  
+
+
+# Our Solution
+
+Our CI/CD solution is illustrated in [Figure 2](#fig2). We don’t explicitly create a directed acyclic graph (DAG)  to connect the steps in an ML workflow (e.g. preprocessing, training, validation, deployment, etc…). Rather, we use a set of independent controllers. Each controller declaratively describes the desired state of the world and takes  actions necessary to make the actual state of the world match. This independence makes it easy for us to use whatever tools make the most sense for each step. More specifically we use
+
+*   Jupyter notebooks for developing models. 
+*   GitOps for continuous integration and deployment. 
+*   Kubernetes and managed cloud services for underlying infrastructure.  
+
+<img id="fig2" src="/images/2020-08-01-data-science-meets-devops/fig2.ci-cd.png" width="" alt="alt_text" title="">
+<figcaption><strong>Figure 2:</strong> illustrates how we do CI/CD. Our pipeline today consists of two independently operating controllers. We configure the Trainer (left hand side) by describing what models we want to exist; i.e. what it means for our models to be “fresh”.  The Trainer periodically checks whether the set of trained models are sufficiently fresh and if not trains a new model. We likewise configure the Deployer (right hand side) to define what it means for the deployed model to be in sync with the set of trained models. If the correct model is not deployed it will deploy a new model.</figcaption>
+
+For more details on model training and deployment refer to the [Actuation section below](#actuation).
+
+# Background
+
+## Building Resilient Systems With Reconcilers
+
+A reconciler is a control pattern that has proven to be immensely useful for building resilient systems. The reconcile pattern is [at the heart of how Kubernetes works](https://book.kubebuilder.io/cronjob-tutorial/controller-overview.html). Figure 3 illustrates how a reconciler works. A reconciler works by first observing the state of the world; e.g. what model is currently deployed. The reconciler then compares this against the desired state of the world and computes the diff; e.g the model with label “version=20200724” should be deployed, but the model currently deployed has label “version=20200700”. The reconciler then takes the action necessary to drive the world to the desired state; e.g. open a pull request to change the deployed model.
+
+
+<img src="/images/2020-08-01-data-science-meets-devops/fig3.reconciler.png" width="" alt="Figure 3" title="Figure 3. Illustration of the reconciler pattern as applied by our deployer.">
```

Comment:
Do we want to make Figure 3 `strong`, like in other image names?

Reply:
Ill change this now